lib/Dialect/ESI/runtime/cosim/lib/RpcServer.cpp

```diff
@@ -0,0 +1,355 @@
+  std::cout << "Server listening on 127.0.0.1:" << port << std::endl;
+}
+
+void Impl::stop() {
+  /// Shutdown the server and wait for it to finish.
+  server->Shutdown();
+  server->Wait();
+  server = nullptr;
+}
+
+Impl::~Impl() {
+  if (server)
+    stop();
+  for (auto &port : readPorts)
+    delete port.second;
+  for (auto &port : writePorts)
+    delete port.second;
+}
+
+ReadChannelPort &Impl::registerReadPort(const std::string &name,
+                                        const std::string &type) {
+  auto port = new RpcServerReadPort(new Type(type));
+  readPorts.emplace(name, port);
+  return *port;
+}
+WriteChannelPort &Impl::registerWritePort(const std::string &name,
+                                          const std::string &type) {
+  auto port = new RpcServerWritePort(new Type(type));
+  writePorts.emplace(name, port);
+  return *port;
+}
+
+ServerUnaryReactor *Impl::GetManifest(CallbackServerContext *context,
+                                      const VoidMessage *, Manifest *response) {
+  response->set_esi_version(esiVersion);
+  response->set_compressed_manifest(compressedManifest.data(),
+                                    compressedManifest.size());
+  ServerUnaryReactor *reactor = context->DefaultReactor();
+  reactor->Finish(Status::OK);
+  return reactor;
+}
+
+/// Load the list of channels into the response and fire it off.
+ServerUnaryReactor *Impl::ListChannels(CallbackServerContext *context,
+                                       const VoidMessage *,
+                                       ListOfChannels *channelsOut) {
+  for (auto [name, port] : readPorts) {
+    auto *channel = channelsOut->add_channels();
+    channel->set_name(name);
+    channel->set_type(port->getType()->getID());
+    channel->set_dir(ChannelDesc::Direction::ChannelDesc_Direction_TO_SERVER);
+  }
+  for (auto [name, port] : writePorts) {
+    auto *channel = channelsOut->add_channels();
+    channel->set_name(name);
+    channel->set_type(port->getType()->getID());
+    channel->set_dir(ChannelDesc::Direction::ChannelDesc_Direction_TO_CLIENT);
+  }
+
+  // The default reactor is basically to just finish the RPC call as if we're
+  // implementing the RPC function as a blocking call.
+  auto reactor = context->DefaultReactor();
+  reactor->Finish(Status::OK);
+  return reactor;
+}
+
+namespace {
+/// When a client connects to a read port (on its end, a write port on this
+/// end), construct one of these to poll the corresponding write port on this
+/// side and forward the messages.
+class RpcServerWriteReactor : public ServerWriteReactor<esi::cosim::Message> {
+public:
+  RpcServerWriteReactor(RpcServerWritePort *writePort)
+      : writePort(writePort), sentSuccessfully(SendStatus::UnknownStatus),
+        shutdown(false) {
+    myThread = std::thread(&RpcServerWriteReactor::threadLoop, this);
+  }
+  ~RpcServerWriteReactor() {
+    shutdown = true;
+    myThread.join();
+  }
+
+  void OnDone() override { delete this; }
```

Comment:
What's the argument for doing it like this, as opposed to more traditional (re: safe) memory management techniques?
I can tell that this is connected to the logic in Impl::ConnectToClientChannel. In the spirit of measureability and general awareness of what's going on in the system, it seems counterintuitive to spawn off these unmanaged objects out into the eather, with no ties (re: management) from the outside world. I'd rather see a design where such write reactors signal back to their parent management objects that they have finished doing whatever they need to do. Their managers can then react accordingly (log, trace, reorder, do nothing, ...) and safely destruct the reactor.

Reply:
Yeah, this is definitely an odd paradigm. It's copied from one of the gRPC examples: https://github.com/grpc/grpc/blob/4795c5e69b25e8c767b498bea784da0ef8c96fd5/examples/cpp/route_guide/route_guide_callback_server.cc#L120 so I assume it's the only way to know that gRPC has release the reference to this object. (I'll add a note to that effect.)

This model is equivalent letting gRPC do the memory management of this object, so it's not unmanaged or thrown into the ether. My gut actually agrees with yours, but I can't think of a tangible reason to add that complexity. In terms of measurability, what thread deletes the code isn't terribly relevant as the object could contain a reference to `Impl` or whatever is responsible for measurement, logging, or something else. If the management class wants to delete this object before gRPC is done with it, it can't -- that would be unsafe since gRPC isn't done with it.

I think this is the right way to go for now, and can be changed later on if there's a specific reason. There's no reason to prematurely add complexity.